### DISCUSSION TITLE
(Lack of) consistency of docstring quotation marks

### DISCUSSION THREAD

**Discussion by arporter** [stfc/PSyclone#2797] [General]
@schreiberx has pointed out that we're not consistent in whether we use ''' or """ for docstrings. In the wiki example markup we use ''' (but don't say that's required) and, in fparser, Black has converted everything to """. I would therefore vote that we madate the use of """, update the wiki appropriately, and update docstrings on a "you touch it, you fix it" basis. (Or we find a quiet slot over Christmas and do it as a big bang.)
What do @TeranIvy @hiker and @sergisiso think?

**Comment by schreiberx** [comment]
You didn't ask me, but I like how this is handled in fparser: Just run `black` over it and that's it.

**Comment by arporter** [comment]
Oh yes, that's a good point. Much though I'm not totally a fan of black, having one less thing to worry about when doing a code review is very valuable.

**Reply by schreiberx** [reply]
But please buy a new laptop with a larger screen and allow 128 characters per line :-)

**Reply by arporter** [reply]
I don't want a larger laptop - it's me who has to carry it :-) Maybe once we have flexible screens that we can unroll/unfold or something then I might consider it! Or maybe I just use VR and then I can have as big a screen as I want?

**Reply by schreiberx** [reply]
Or you can make the font size smaller and buy some magnifying glasses  :-)

**Reply by sergisiso** [reply]
I agree in letting black (or ruff) doing the formatting. Now is a bit inconvenient for me as I have big unmerged branches with many conflicts, but we could agree a point in the future to do the transition.

Regarding line length, 128 is too long, PEP8 recommends 79 but we could move to the default 88 that black and ruff have. Using defaults is good.

**Reply by hiker** [reply]
Agree with Sergi - esp. the 'now is a bit inconvenient'

**Reply by schreiberx** [reply]
Would you be able to make this for the PSyclone 3.0 release?
If you make this change, I suggest switching also from sphinx-type information to Python-type annotations:
https://pypi.org/project/sphinx-autodoc-typehints/

**Reply by hiker** [reply]
IMHO no way for 3.0. And still the case, I have two long-outstanding PRs that would potentially very heavy affected by any global change (due to refactoring code from lfric and gocean into a common base class), causing me a lot of overhead. Also while I am in favour of types, if we require full type checking compliance ny running e.g. `myp`, expect quite a bit of overhead, python typing sometimes behaves quite unexpected (not saying wrong ;) ), esp. when using optional variables.

**Reply by sergisiso** [reply]
We decided with @arporter that for type-annotations we won't do a big bang change for now, instead we will follow a gradual approach. We are allowed/encouraged to start using them for new code or code being changed, if they are supported by our oldest python in the tests (currently 3.7). It is ok to still use sphinx-types for cases that are not well supported with this older python versions.

We saw that Sphinx already handles them with linkable references in the autodocs by default, what does the shpinx-autodoc-typehints do @schreiberx ? Just provide more control on how these are formatted?

**Reply by schreiberx** [reply]
What do you mean by "what does the shpinx-autodoc-typehints"? It supports type hints in sphinx but that was probably not your question.
The advantage of type hints is that they can be used by modern code development tools to provide more information on the variable, do additional type-checking in the editor, etc.

**Reply by sergisiso** [reply]
Type-hints in sphinx already work out-of-the-box without this extension, that's why I am asking

**Reply by schreiberx** [reply]
Let's please discuss this f2f on Friday. I have the impression we're both assuming different things as to why we should use type hints directly in Python.
